lib/websocket/websocket.js

```diff
@@ -57,38 +58,48 @@ class WebSocket extends EventTarget {
     url = webidl.converters.USVString(url)
     protocols = options.protocols
 
-    // 1. Let urlRecord be the result of applying the URL parser to url.
+    // 1. Let baseURL be this's relevant settings object's API base URL.
+    const baseURL = getGlobalOrigin()
+
+    // 1. Let urlRecord be the result of applying the URL parser to url with baseURL.
     let urlRecord
 
     try {
-      urlRecord = new URL(url)
+      urlRecord = new URL(url, baseURL)
     } catch (e) {
-      // 2. If urlRecord is failure, then throw a "SyntaxError" DOMException.
+      // 3. If urlRecord is failure, then throw a "SyntaxError" DOMException.
       throw new DOMException(e, 'SyntaxError')
     }
 
-    // 3. If urlRecord’s scheme is not "ws" or "wss", then throw a
-    //    "SyntaxError" DOMException.
+    // 4. If urlRecord’s scheme is "http", then set urlRecord’s scheme to "ws".
+    if (urlRecord.protocol === 'http:') {
+      urlRecord.protocol = 'ws:'
+    } else if (urlRecord.protocol === 'https:') {
+      // 5. Otherwise, if urlRecord’s scheme is "https", set urlRecord’s scheme to "wss".
+      urlRecord.protocol = 'wss:'
+    }
+
+    // 6. If urlRecord’s scheme is not "ws" or "wss", then throw a "SyntaxError" DOMException.
     if (urlRecord.protocol !== 'ws:' && urlRecord.protocol !== 'wss:') {
       throw new DOMException(
         `Expected a ws: or wss: protocol, got ${urlRecord.protocol}`,
         'SyntaxError'
       )
     }
 
-    // 4. If urlRecord’s fragment is non-null, then throw a "SyntaxError"
+    // 7. If urlRecord’s fragment is non-null, then throw a "SyntaxError"
     //    DOMException.
-    if (urlRecord.hash) {
+    if (urlRecord.hash || urlRecord.href.endsWith('#')) {
```

Comment:
I think we should have a method in the WHATWG url to get if the fragment is `non-null`.

Nit: `urlRecord.hash.length > 0 || urlRecord.href.endsWith('#')`